conda/gateways/connection/adapters/jlap/adapter.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Cached repodata with `.jlap` incremental diffs, implemented as a ConnectionAdapter.
+"""
+
+import gzip
+import json
+import logging
+
+from requests import PreparedRequest, Response
+from requests.adapters import HTTPAdapter
+
+from conda.base import constants as consts
+from .repodata_proxy import RepodataUrl, fetch_repodata_json, static_file_headers, patch_files
+from .sync_jlap import SyncJlap
+
+log = logging.getLogger(__name__)
+
+
+class JlapAdapter(HTTPAdapter):
+    def send(self, request, **kwargs):
+        """
+        Overrides the default HTTPAdapter. When this particular HTTPAdapter is in
+        use, we only care about intercepting URLs that end with
+        ``conda.base.constants.REPODATA_FN``
+        """
+        if request.url.endswith(consts.REPODATA_FN):
+            log.debug("Intercept %s", request.url)
+
+            # Pre-processing to the request, including changing its URL
+            repodata_url = RepodataUrl(request.url)
+            sync = SyncJlap(repodata_url)
+            new_request = self.prepare_request(request, sync)
+            response = self.send(new_request)
+
+            # This means there was an error with the response. We return it immediately.
+            if response.status_code > 300:
+                return response
+
+            # This step does some saving to the filesystem plus other processing
+            response = self.post_response_processing(response, sync)
+
+            # Post-processing to the response, including changing back its URL
+            return self.prepare_response(response, request)
+        else:
+            log.debug("Skip intercept %s", request.url)
+            return super().send(request, **kwargs)
+
+    @staticmethod
+    def prepare_request(request: PreparedRequest, sync: SyncJlap) -> PreparedRequest:
+        """
+        We need to specially prepare the request because it's actually being diverted
+        to a different location.
+
+        For example:
+            "https://repo.anaconda.com/pkgs/main/noarch/repodata.json"
+
+        Translates to:
+            "https://repodata.flydev.io/pkgs/main/noarch/repodata.jlap"
+        """
+        new_headers = sync.get_request_headers()
+        request.url = sync.repodata_url.translate_to_jlap_url()
+        request.headers.update(new_headers)
+
+        return request
+
+    @staticmethod
+    def prepare_response(response: Response, request: PreparedRequest) -> Response:
+        """
+        We do this to make it look like the client is getting back the original thing they
+        requested and not what we actually retrieved (i.e. the JLAP stuff)
+        """
+        response.request = request
+        response.url = request.url
+        # response.status_code = 200  TODO: Is this really necessary? Need to ask dholth
```

Comment:
it might be less necessary with bottle removed, if it runs without throwing an exception, it's probably fine?